### DISCUSSION TITLE
Can you read on serial and send on usb?

### DISCUSSION THREAD

**Discussion by fapplin** [FortySevenEffects/arduino_midi_library#336] [Q&A] [answered]
OK. With the changes I'm reading about - here's what I'm trying to do? I'm using a Raspberry Pi Pico and Earle Philhower's Arduino core. 

I need to read from a 5 DIN Serial (with RX and TX pins defined) and if possible use callbacks. In the callbacks, I will be changing notes based on certain criteria and then sending the new/changed notes out via USB connected to a computer (hopefully within the callbacks using sendNoteOn and sendNoteOff).

Is this possible using only this library? 

Currently, I am reading midi data via the serial port (pins 16, 17 on the pico) and parsing the midi data myself. I'm using your USB capability via 

```
// USB MIDI object
Adafruit_USBD_MIDI usb_midi;

// Create a new instance of the Arduino MIDI Library,
// and attach usb_midi as the transport.
MIDI_CREATE_INSTANCE(Adafruit_USBD_MIDI, usb_midi, MIDI);
```

and using MIDI to do the sendNoteOn and sendNoteOff. I need all other midi messages to pass to just pass through (which I was hoping to use your new filter capability). 

So, basically I'm asking can your library do all of this or do I need to just travel the path I've been traveling?

Sorry for all the questions and confusion. 

Frank

**Comment by franky47** [comment]
Yes, if you create two instances of the library, one bound to USB and one to DIN / HardwareSerial, you can register a callback on the DIN instance and call send methods on the USB instance, they are independent.

As a matter of fact, this is what the [DualMerger example](https://github.com/FortySevenEffects/arduino_midi_library/blob/master/examples/DualMerger/DualMerger.ino) does.

**Reply by fapplin** [reply]
OK. A Pico has 2 serial ports. I have one (the default) in use accepting commands from a touch screen device. So, I have to use the other serial port for the midi, which I do on the RX/TX on pins 16 and 17.  Does your serial transport allow me to define the pins of where its receiving and transmitting?

And, truly, thank you for all of your help.

**Reply by franky47** [reply]
Not the pins directly, but you should have a HardwareSerial instance associated with those pins, and then it's just a matter of telling the library which one to use (the DualMerger example also shows how to do this with the `MIDI_CREATE_INSTANCE` macro).

**Comment by fapplin** [comment]
OK. I have the sendNoteOn() and sendNoteOff() passing between the serial and the USB!! Soooo, much easier. But, now, come the filter part.

This was your proposed filter from a discussion in August.

```
bool everythingButNotes(const midi::Message& message)
{
  return message.type != midi::NoteOn && message.type != midi::NoteOff;
}

// The usual callbacks to handle Note messages:
void onNoteOn(byte pitch, byte velocity, byte channel)
{
}

void onNoteOff(byte pitch, byte velocity, byte channel)
{
}

void setup()
{
  MIDI.setThruFilter(everythingButNotes);
  MIDI.setHandleNoteOn(onNoteOn);
  MIDI.setHandleNoteOff(onNoteOff);
  // ...
}
```

That doesn't work. Has this filter part been implemented?

**Reply by franky47** [reply]
It's not yet available in the last release, only here on GitHub. You'll have to [download the code from the `v5.1.0` branch](https://github.com/FortySevenEffects/arduino_midi_library/tree/feat/v5.1.0/src) and replace it in your project to gain access to the filter/map feature.

**Reply by fapplin** [reply]
OK. I'll check it out! Thanks for all the help!

**Reply by fapplin** [reply]
OK. I replaced the files with the files you pointed to from the v5.1.0 branch and I used your example that I cited above. I get the following error.

```
KISIMIDI_101123:22:25: error: template placeholder type 'const Message<...auto...>' must be followed by a simple declarator-id
   22 | bool everythingButNotes(const midi::Message& message)
      |                         ^~~~~
In file included from /home/frank/Arduino/libraries/MIDI_Library/src/MIDI.h:33,
                 from /home/frank/KISIMIDI_101123/KISIMIDI_101123.ino:5:
/home/frank/Arduino/libraries/MIDI_Library/src/midi_Message.h:42:8: note: 'template<unsigned int SysExMaxSize> struct midi::Message' declared here
   42 | struct Message
      |        ^~~~~~~
KISIMIDI_101123:22:25: error: template placeholder type 'const Message<...auto...>' must be followed by a simple declarator-id
   22 | bool everythingButNotes(const midi::Message& message)
      |                         ^~~~~
In file included from /home/frank/Arduino/libraries/MIDI_Library/src/MIDI.h:33,
                 from /home/frank/KISIMIDI_101123/KISIMIDI_101123.ino:5:
/home/frank/Arduino/libraries/MIDI_Library/src/midi_Message.h:42:8: note: 'template<unsigned int SysExMaxSize> struct midi::Message' declared here
   42 | struct Message
      |        ^~~~~~~
/home/frank/KISIMIDI_101123/KISIMIDI_101123.ino: In function 'bool everythingButNotes(...)':
KISIMIDI_101123:24:10: error: 'message' was not declared in this scope
   24 |   return message.type != midi::NoteOn && message.type != midi::NoteOff;
      |          ^~~~~~~
/home/frank/KISIMIDI_101123/KISIMIDI_101123.ino: In function 'void setup()':
KISIMIDI_101123:105:28: error: invalid conversion from 'bool (*)(...)' to 'midi::MidiInterface<midi::SerialMIDI<arduino::HardwareSerial> >::ThruFilterCallback' {aka 'bool (*)(const midi::Message<128>&)'} [-fpermissive]
  105 |   midiSerial.setThruFilter(everythingButNotes);
      |                            ^~~~~~~~~~~~~~~~~~
      |                            |
      |                            bool (*)(...)
In file included from /home/frank/KISIMIDI_101123/KISIMIDI_101123.ino:5:
/home/frank/Arduino/libraries/MIDI_Library/src/MIDI.h:244:60: note:   initializing argument 1 of 'midi::MidiInterface<Transport, _Settings, _Platform>& midi::MidiInterface<Transport, _Settings, _Platform>::setThruFilter(midi::MidiInterface<Transport, _Settings, _Platform>::ThruFilterCallback) [with Transport = midi::SerialMIDI<arduino::HardwareSerial>; _Settings = midi::DefaultSettings; _Platform = midi::DefaultPlatform; midi::MidiInterface<Transport, _Settings, _Platform>::ThruFilterCallback = bool (*)(const midi::Message<128>&); midi::MidiInterface<Transport, _Settings, _Platform>::MidiMessage = midi::Message<128>]'
  244 |     inline MidiInterface& setThruFilter(ThruFilterCallback fptr)
      |                                         ~~~~~~~~~~~~~~~~~~~^~~~
Using library Adafruit_TinyUSB_Arduino at version 1.13.3 in folder: /home/frank/Arduino/hardware/pico/rp2040/libraries/Adafruit_TinyUSB_Arduino 
Using library MIDI_Library at version 5.0.2 in folder: /home/frank/Arduino/libraries/MIDI_Library 
exit status 1
template placeholder type 'const Message<...auto...>' must be followed by a simple declarator-id
```

**Reply by franky47** [reply]
Ah yeah the Message struct is template over the sysex size, try this:

```c++
// Set the template argument to the SysEx buffer size in settings (default 128):
bool everythingButNotes(const midi::Message<128>& message)
{
  return message.type != midi::NoteOn && message.type != midi::NoteOff;
}
```

**Reply by fapplin** [reply]
OK. It compiled. So, should this handle all other messages like pitch bend? Or, do I still code for that? I commented my pitch bend callback out to see if this handled it, but it doesn't seem to.

**Reply by franky47** [reply]
This should forward all incoming message types but NoteOn and NoteOff to the associated output (if it supports Thru, this would only work on serial transports).

**Reply by fapplin** [reply]
OK. The incoming midi is serial, but outgoing is USB. So, I guess the filter doesn't buy me much. Getting the callbacks and just doing a simple midiSerial.read() in the loop() without the need to parse anything is great, though. Saved a hundred or so lines of code!

**Reply by franky47** [reply, marked as answer]
Ah, indeed this filter/map thing only works for Thru on the **same** interface, but there's a way to Thru on another interface, by hacking the filter callback and sending from there:

```c++
bool everythingButNotes(const midi::Message<128>& message)
{
  const bool forward = message.type != midi::NoteOn && message.type != midi::NoteOff;
  if (forward)
  {
    usbMidi.send(message);
  }
  // Thru back to the Serial output
  return forward;
}
```

**Comment by fapplin** [comment]
That seemed to do it. I commented out the following:

```
//  midiSerial.setHandlePitchBend(handlePitchBend);
//  midiSerial.setHandleAfterTouchPoly(handleAfterTouchPoly);
//  midiSerial.setHandleAfterTouchChannel(handleAfterTouchChannel);
//  midiSerial.setHandleControlChange(handleControlChange);
//  midiSerial.setHandleProgramChange(handleProgramChange);
//  midiSerial.setHandleSystemExclusive(handleSystemExclusive);
//  midiSerial.setHandleSystemReset(handleSystemReset);
```

And, I know at least the pitch bend and mod wheel are working with just the filter! Cool! Thanks.
